docs/developers/talawa/flutter_testing_guide.md

```diff
@@ -0,0 +1,271 @@
+var myMock = MockMyObject();
+when(myMock.myMethod()).thenReturn('my result');
+```
+This will replace the `myMethod()` method on the `myMock` object with a stub that always returns the string 'my result'.
+
+`when()` method is used in the previous Calculator example as well where it is used to stub the add method to mock the Calculator.
+
+You can also use the any matcher to match any input value. For example:
+
+```dart
+when(myMock.myMethod(any)).thenReturn('my result');
+```
+This will stub the `myMethod()` method to always return 'my result', regardless of the input value.
+
+## Talawa App Tests Example on How to use Mocks and Stubs Correctly
+
+There is a function in talawa which provides chat service in the app. The file is located in `talawa/lib/services/chat_service.dart`. Unit Tests for this file have been written to test all the methods written in the file for instance `sendMessageToDirectChat` and `getDirectChatsByUserId` methods. These tests are written in the file `talawa\test\service_tests\chat_service_test.dart`. These tests have all the necessary mocks and stubs to understand the concept of testing in talawa.
+
+### sendMessageToDirectChat Method
+
+```dart
+Future<void> sendMessageToDirectChat(
+    String chatId,
+    String messageContent,
+  ) async {
+    // trigger graphQL mutation to push the message in the Database.
+    final result = await _dbFunctions.gqlAuthMutation(
+      ChatQueries().sendMessageToDirectChat(),
+      variables: {"chatId": chatId, "messageContent": messageContent},
+    );
+
+    final message = ChatMessage.fromJson(
+      result.data['sendMessageToDirectChat'] as Map<String, dynamic>,
+    );
+
+    _chatMessageController.add(message);
+
+    debugPrint(result.data.toString());
+  }
+```
+
+Test written for this method looks like this
+
+```dart
+test('Test SendMessageToDirectChat Method', () async {
+      final dataBaseMutationFunctions = locator<DataBaseMutationFunctions>();
+      const id = "1";
+      const messageContent = "test";
+
+      final query = ChatQueries().sendMessageToDirectChat();
+      when(
+        dataBaseMutationFunctions.gqlAuthMutation(
+          query,
+          variables: {
+            "chatId": id,
+            "messageContent": messageContent,
+          },
+        ),
+      ).thenAnswer(
+        (_) async => QueryResult(
+          options: QueryOptions(document: gql(query)),
+          data: {
+            'sendMessageToDirectChat': {
+              '_id': id,
+              'messageContent': messageContent,
+              'sender': {
+                'firstName': 'Mohamed',
+              },
+              'receiver': {
+                'firstName': 'Ali',
+              }
+            },
+          },
+          source: QueryResultSource.network,
+        ),
+      );
+      final service = ChatService();
+      await service.sendMessageToDirectChat(
+        id,
+        messageContent,
+      );
+    })
+```
+
+Here is a breakdown of what this test does
```

Comment:
There should be a title here too. The segmentation makes the readability better.

Reply:
What should be the title here?

Reply:
Think of something appropriate given my prior advice